turbinia/api/models/request_status.py

```diff
@@ -0,0 +1,135 @@
+# -*- coding: utf-8 -*-
+# Copyright 2022 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Turbinia API server Request models."""
+
+import datetime
+import logging
+
+from typing import Optional, List, Dict
+from pydantic import BaseModel
+from turbinia import state_manager
+from turbinia import config as turbinia_config
+
+log = logging.getLogger('turbinia:api_server:routes:request')
+
+
+class RequestStatus(BaseModel):
+  """Represents a Turbinia request status object."""
+  request_id: str = None
+  tasks: List[Dict] = []
+  reason: str = None
+  requester: str = None
+  last_task_update_time: str = None
+  status: str = None
+  task_count: int = 0
+  successful_tasks: int = 0
+  running_tasks: int = 0
+  failed_tasks: int = 0
+
+  def get_request_data(
+      self, request_id: str, tasks: Optional[List[Dict]] = None,
+      summary: bool = False):
+    """Gets task information for a specific Turbinia request.
+
+    Args:
+      request_id (str): A Turbinia request identifier.
+      tasks (Optional[List[Dict]]): an optional list of task objects.
+      summary (bool): If this flag is set, the 'tasks' list will be empty.
+
+    Returns:
+      bool: True if the request has at least one task associated with it.
+    """
+    if not tasks:
+      _state_manager = state_manager.get_state_manager()
+      tasks = _state_manager.get_task_data(
+          instance=turbinia_config.INSTANCE_ID, request_id=request_id)
+
+    if not summary:
+      for task in tasks:
+        current_request_id = task.get('request_id')
+        if current_request_id == request_id:
+          self.tasks.append(task)
+
+    for task in tasks:
+      self.request_id = task.get('request_id')
+      self.requester = task.get('requester')
+      self.reason = task.get('reason')
+      self.task_count = len(tasks)
+      task_status = task.get('status')
+      if isinstance(task.get('last_update'), datetime.datetime):
+        task_last_update = datetime.datetime.timestamp(task.get('last_update'))
+      else:
+        task_last_update = task.get('last_update')
+
+      if not self.last_task_update_time:
+        self.last_task_update_time = task_last_update
+      else:
+        self.last_task_update_time = max(
+            self.last_task_update_time, task_last_update)
+      if task.get('successful'):
+        self.successful_tasks += 1
+      elif task_status:
+        if 'running' in task_status:
+          self.running_tasks += 1
+      else:
```

Comment:
Done

Reply:
As discussed, this may also need to be addressed in the appropriate class to disambiguate running/queued statuses and save the proper status in Redis.